diff -Naur a/package/Config.in b/package/Config.in
--- a/package/Config.in	2020-06-24 01:44:58.314989000 +0000
+++ b/package/Config.in	2020-06-23 21:12:34.068101000 +0000
@@ -1219,6 +1219,7 @@
 	source "package/python-zope-interface/Config.in"
 endmenu
 endif
+	source "package/qjs/Config.in"
 	source "package/ruby/Config.in"
 	source "package/tcl/Config.in"
 if BR2_PACKAGE_TCL
diff -Naur a/package/qjs/01-add-libatomic.patch b/package/qjs/01-add-libatomic.patch
--- a/package/qjs/01-add-libatomic.patch	1970-01-01 00:00:00.000000000 +0000
+++ b/package/qjs/01-add-libatomic.patch	2020-06-23 21:50:34.686555000 +0000
@@ -0,0 +1,12 @@
+diff -Naur quickjs-2020-04-12.old/Makefile quickjs-2020-04-12.new/Makefile
+--- quickjs-2020-04-12.old/Makefile	2020-04-12 13:18:13.000000000 +0000
++++ quickjs-2020-04-12.new/Makefile	2020-06-23 21:50:13.452255000 +0000
+@@ -166,7 +166,7 @@
+ QJS_OBJS+=$(OBJDIR)/qjscalc.o
+ endif
+
+-LIBS=-lm
++LIBS=-lm -latomic
+ ifndef CONFIG_WIN32
+ LIBS+=-ldl
+ endif
diff -Naur a/package/qjs/Config.in b/package/qjs/Config.in
--- a/package/qjs/Config.in	1970-01-01 00:00:00.000000000 +0000
+++ b/package/qjs/Config.in	2020-06-23 21:13:36.001192000 +0000
@@ -0,0 +1,6 @@
+config BR2_PACKAGE_QJS
+	bool "qjs"
+	help
+	  QuickJS Javascript Engine
+
+	  https://bellard.org/quickjs/
diff -Naur a/package/qjs/qjs.hash b/package/qjs/qjs.hash
--- a/package/qjs/qjs.hash	1970-01-01 00:00:00.000000000 +0000
+++ b/package/qjs/qjs.hash	2020-06-24 01:47:08.772021000 +0000
@@ -0,0 +1 @@
+sha1  5fc6872e8cf307b32dd8245ca96949acde77b1ba  quickjs-2020-04-12.tar.xz
diff -Naur a/package/qjs/qjs.mk b/package/qjs/qjs.mk
--- a/package/qjs/qjs.mk	1970-01-01 00:00:00.000000000 +0000
+++ b/package/qjs/qjs.mk	2020-06-24 01:39:20.635123000 +0000
@@ -0,0 +1,21 @@
+################################################################################
+#
+# qjs
+#
+################################################################################
+
+QJS_VERSION = 2020-04-12
+QJS_SOURCE = quickjs-$(QJS_VERSION).tar.xz
+QJS_SITE = https://bellard.org/quickjs
+QJS_LICENSE = MIT
+
+define QJS_BUILD_CMDS
+    echo $(TARGET_CONFIGURE_OPTS)
+    $(MAKE) CROSS_PREFIX=$(TARGET_CROSS) -C $(@D)
+endef
+
+define QJS_INSTALL_TARGET_CMDS
+    $(INSTALL) -D -m 0755 $(@D)/qjs $(TARGET_DIR)/usr/bin
+endef
+
+$(eval $(generic-package))
